crypto3/libs/hash/include/nil/crypto3/hash/detail/merkle_damgard_construction.hpp

```diff
@@ -151,12 +155,15 @@ namespace nil {
                     using namespace nil::crypto3::detail;
 
                     std::array<length_type, 1> length_array = {{length}};
-                    std::array<word_type, length_words> length_words_array;
-                    pack<endian_type, endian_type, length_bits, word_bits>(length_array.begin(), length_array.end(),
-                                                                           length_words_array.begin());
-                    // Append length
-                    for (std::size_t i = length_words; i; --i)
-                        block[block_words - i] = length_words_array[length_words - i];
+                    std::array<word_type, length_type_bits / word_bits> length_words_array;
+                    pack<endian_type, endian_type, length_type_bits, word_bits>(
+                        length_array.begin(), length_array.end(),
+                        length_words_array.begin());
+
+                    // Append length, but from the end. We were required to write length in 'length_bits' bits,
+                    // but actually used just 'length_type_bits' bits.
+                    for (int i = length_type_bits / word_bits; i > 0; --i)
+                        block[block_words - i] = length_words_array[length_type_bits / word_bits - i];
```

Comment:
```suggestion
                    for (int i = length_words; i > 0; --i)
                        block[block_words - i] = length_words_array[length_words - i];
```

Reply:
Same here, length_words is the total length we MUST use, not the actually length used. We want the rest of it to stay 0.

Reply:
My bad, missed that it is `length_type_bits`, not `length_bits`